utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java

```diff
@@ -0,0 +1,307 @@
+
+            if (!utilityElementTraceResult.getElements().isEmpty()) {
+
+              // iterate through the map's feature layers
+              mapView.getMap().getOperationalLayers().forEach(layer -> {
+                if (layer instanceof FeatureLayer) {
+
+                  // create query parameters to find features whose network source name matches the layer's feature
+                  // table name
+                  QueryParameters queryParameters = new QueryParameters();
+                  utilityElementTraceResult.getElements().forEach(utilityElement -> {
+
+                    String networkSourceName = utilityElement.getNetworkSource().getName();
+                    String featureTableName = ((FeatureLayer) layer).getFeatureTable().getTableName();
+
+                    if (networkSourceName.equals(featureTableName)) {
+                      queryParameters.getObjectIds().add(utilityElement.getObjectId());
+                    }
+                  });
+
+                  // select features that match the query
+                  ListenableFuture<FeatureQueryResult> featureQueryResultListenableFuture =
+                          ((FeatureLayer) layer).selectFeaturesAsync(queryParameters, FeatureLayer.SelectionMode.NEW);
+
+                  // wait for the selection to finish
+                  featureQueryResultListenableFuture.addDoneListener(() -> {
+                    // update the status text, enable the buttons and hide the progress indicator
+                    statusLabel.setText("Isolation trace completed.");
+                    enableUI(true);
+                  });
+                }
+              });
+
+            } else {
+              statusLabel.setText("Isolation trace completed.");
+              new Alert(Alert.AlertType.INFORMATION, "Isolation trace returned no elements.").show();
+              enableUI(true);
+            }
+
+          } else {
+            statusLabel.setText("Trace failed.");
+            new Alert(Alert.AlertType.ERROR, "Isolation trace result is not a utility element.").show();
+            enableUI(true);
+          }
+
+        } catch (Exception e) {
+          statusLabel.setText("Trace failed.");
+          new Alert(Alert.AlertType.ERROR, "Error getting isolation trace result.").show();
+          enableUI(true);
+        }
+      });
+
+    } catch (Exception e) {
+      new Alert(Alert.AlertType.ERROR, "Error performing isolation trace.").show();
+      enableUI(true);
+    }
+  }
+
+
+  /**
+   * Enables/disables the UI and hides/shows the progress indicator.
+   *
+   * @param enable whether to enable or disable the UI
+   */
+  private void enableUI(boolean enable) {
+    progressIndicator.setVisible(!enable);
+    traceButton.setDisable(enable);
+    categorySelectionComboBox.setDisable(enable);
+    includeIsolatedFeaturesCheckbox.setDisable(enable);
```

Comment:
I find that `!enable` needs to be passed to all 4 lines above

Reply:
Thanks for spotting that! I think the delay during the traces was quite short so I've never really noticed this misbehaving.